packages/adapter-node/src/index.js

```diff
@@ -6,10 +6,90 @@ export const path = env('SOCKET_PATH', false);
 export const host = env('HOST', '0.0.0.0');
 export const port = env('PORT', !path && '3000');
 
+const shutdown_timeout = parseInt(env('SHUTDOWN_TIMEOUT', '30'));
+const idle_timeout = parseInt(env('IDLE_TIMEOUT', '0'));
+const listen_pid = parseInt(env('LISTEN_PID', '0'));
+const listen_fds = parseInt(env('LISTEN_FDS', '0'));
+// https://www.freedesktop.org/software/systemd/man/latest/sd_listen_fds.html
+const SD_LISTEN_FDS_START = 3;
+
+if (listen_pid !== 0 && listen_pid !== process.pid) {
+	throw new Error(`received LISTEN_PID ${listen_pid} but current process id is ${process.pid}`);
+}
+if (listen_fds > 1) {
+	throw new Error(
+		`only one socket is allowed for socket activation, but LISTEN_FDS was set to ${listen_fds}`
+	);
+}
+
+const socket_activation = listen_pid === process.pid && listen_fds === 1;
+
+let requests = 0;
+/** @type {NodeJS.Timeout | void} */
+let shutdown_timeout_id;
+/** @type {NodeJS.Timeout | void} */
+let idle_timeout_id;
+
 const server = polka().use(handler);
 
-server.listen({ path, host, port }, () => {
-	console.log(`Listening on ${path ? path : host + ':' + port}`);
-});
+function shutdown() {
+	if (shutdown_timeout_id) return;
+
+	// @ts-expect-error this was added in 18.2.0 but is not reflected in the types
+	server.server.closeIdleConnections();
+
+	server.server.close(() => {
+		if (shutdown_timeout_id) {
+			shutdown_timeout_id = clearTimeout(shutdown_timeout_id);
+		}
+		if (idle_timeout_id) {
+			idle_timeout_id = clearTimeout(idle_timeout_id);
+		}
+	});
+
+	shutdown_timeout_id = setTimeout(
+		// @ts-expect-error this was added in 18.2.0 but is not reflected in the types
+		() => server.server.closeAllConnections(),
+		shutdown_timeout * 1000
+	);
+}
+
+server.server.on(
+	'request',
+	/** @param {import('node:http').IncomingMessage} req */
+	(req) => {
+		requests++;
+
+		if (socket_activation && idle_timeout_id) {
+			idle_timeout_id = clearTimeout(idle_timeout_id);
+		}
+
+		req.on('close', () => {
+			requests--;
+
+			if (shutdown_timeout_id) {
```

Comment:
should this be guarded with requests === 0 too? currently if there are a lot of open requests when shutdown is called, it is going to call closeIdleConnections a lot of times.

And should the server block accepting new requests when shutdown has started?

Reply:
Yeah good point. Not sure why I didn't think about that. I tested the code multiple times with 1000 - 2000 req/sec and never encountered any performance issues even when calling `closeIdleConnections` many times. But I think guarding this with `request === 0` should work too. It's a bit hard to verify because it's difficult to test `keep-alive` connections. But I don't see why this wouldn't work so I'll update it.

`server.close()` which is called inside `shutdown()` will already reject new requests. I verified this too in my tests.